index.js

```diff
@@ -877,67 +865,116 @@ class Watcher extends EventEmitter {
 
-    this._onappendBound = debounceify(this._onappend.bind(this))
+    this._lock = mutexify()
+    this._resolveOnChange = null
 
     this._opening = this._ready().catch(safetyCatch)
   }
 
   async _ready () {
     await this.bee.ready()
-    this.latestDiff = this.bee.version
+    this.current = this.bee.snapshot() // point from which to start watching
     this.opened = true
   }
 
-  async _onappend () {
-    this.running = true
+  [Symbol.asyncIterator] () {
+    return this
+  }
 
-    try {
-      await this._run()
-    } catch (err) {
-      if (!this.closed) this.emit('error', err)
-      this.destroy()
-    } finally {
-      this.running = false
-    }
+  _onappend () {
+    const resolve = this._resolveOnChange
+    this._resolveOnChange = null
+    if (resolve) resolve()
   }
 
-  async _run () {
-    if (this.closed) return
-    if (this.opened === false) await this._opening
+  async _waitForChanges () {
+    if (this.current.version < this.bee.version || this.closed) return
+
+    await new Promise(resolve => {
+      this._resolveOnChange = resolve
+    })
+  }
 
-    const snapshot = this.bee.snapshot()
-    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+  async next () {
+    const release = await this._lock()
 
     try {
-      for await (const data of this.stream) { // eslint-disable-line
-        this.emit('change', snapshot.version, this.latestDiff)
-        break
+      if (this.closed) return { value: undefined, done: true }
+
+      if (!this.opened) await this._opening
+
+      while (true) {
+        await this._waitForChanges()
+
+        if (this.closed) return { value: undefined, done: true }
+
+        if (this.previous) await this.previous.close()
+        this.previous = this.current.snapshot()
+
+        if (this.current) await this.current.close()
+        this.current = this.bee.snapshot()
+
+        this.stream = this.current.createDiffStream(this.previous.version, this.range)
+
+        try {
+          for await (const data of this.stream) { // eslint-disable-line
+            return { done: false, value: { current: this.current, previous: this.previous } }
+          }
+        } finally {
+          this.stream = null
+        }
       }
     } finally {
-      this.stream = null
-      this.latestDiff = snapshot.version
-
-      await snapshot.close()
+      release()
     }
   }
 
-  destroy () {
+  async return () {
+    await this.destroy()
+    return { done: true }
+  }
+
+  throw (err) { // eslint-disable-line n/handle-callback-err
+    // No-op
+  }
+
+  async destroy () {
     if (this.closed) return
     this.closed = true
 
+    this.bee._watchers.delete(this)
+
     if (this.stream && !this.stream.destroying) {
       this.stream.destroy()
     }
 
-    this.bee._watchers.delete(this)
+    this._onappend() // Continue execution being closed
 
-    this.emit('close')
+    const release = await this._lock()
+    release()
+
+    await this._closeAll()
```

Comment:
i think this should be before the lock flush above

Reply:
I think it could potentially throw in the part where we do the two snapshots

Although, we can try-catch the creation of snapshots, and if it's closed then we just return there and otherwise we let the error throw

Reply:
Like this
![image](https://user-images.githubusercontent.com/12686176/228551282-df53d337-75a7-45c0-be22-1b70a63160c6.png)

Wdyt?

Reply:
Or like this
![image](https://user-images.githubusercontent.com/12686176/228551983-a9910d07-6384-4b11-abba-1fe7a6d81b6b.png)

Reply:
Trying to do a more paranoic version about not leaking any snapshot at any point: https://gist.github.com/LuKks/7bf44d82830c6db48adcd73dc03377bb